skymarshal/token/access_token.go

```diff
@@ -0,0 +1,175 @@
+package token
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	"github.com/concourse/concourse/atc/db"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+//go:generate counterfeiter . Generator
+
+type Generator interface {
+	GenerateAccessToken(claims db.Claims) (string, error)
+}
+
+//go:generate counterfeiter . Parser
+
+type Parser interface {
+	ParseExpiry(raw string) (time.Time, error)
+}
+
+//go:generate counterfeiter . ClaimsParser
+
+type ClaimsParser interface {
+	ParseClaims(idToken string) (db.Claims, error)
+}
+
+func StoreAccessToken(
+	logger lager.Logger,
+	handler http.Handler,
+	generator Generator,
+	claimsParser ClaimsParser,
+	accessTokenFactory db.AccessTokenFactory,
+	userFactory db.UserFactory,
+) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/sky/issuer/token" {
+			handler.ServeHTTP(w, r)
+			return
+		}
+		logger := logger.Session("token-request")
+		logger.Debug("start")
+		defer logger.Debug("end")
+
+		rec := httptest.NewRecorder()
+		handler.ServeHTTP(rec, r)
+
+		var body io.Reader
+		defer func() {
+			copyResponseHeaders(w, rec.Result())
+			if body != nil {
+				io.Copy(w, body)
+			}
+		}()
+		if rec.Code < 200 || rec.Code > 299 {
+			body = rec.Body
+			return
+		}
+		var resp struct {
+			AccessToken  string `json:"access_token"`
+			TokenType    string `json:"token_type"`
+			ExpiresIn    int    `json:"expires_in"`
+			RefreshToken string `json:"refresh_token,omitempty"`
+			IDToken      string `json:"id_token"`
+		}
+		err := json.Unmarshal(rec.Body.Bytes(), &resp)
+		if err != nil {
+			logger.Error("unmarshal-response-from-dex", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		claims, err := claimsParser.ParseClaims(resp.IDToken)
+		if err != nil {
+			logger.Error("parse-id-token", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		resp.AccessToken, err = generator.GenerateAccessToken(claims)
+		if err != nil {
+			logger.Error("generate-access-token", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		err = accessTokenFactory.CreateAccessToken(resp.AccessToken, claims)
+		if err != nil {
+			logger.Error("create-access-token-in-db", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		err = userFactory.CreateOrUpdateUser(claims.Username, claims.Connector, claims.Subject)
+		if err != nil {
+			logger.Error("create-or-update-user", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		newResp, err := json.Marshal(resp)
+		if err != nil {
+			logger.Error("marshal-new-response", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		body = bytes.NewReader(newResp)
+	})
+}
```

Comment:
I am not clear end-to-end on what happens if I happen to unilaterally POST to `/sky/issuer/token` with a body of my choosing - perhaps containing all sorts of claims. I haven't tested this PR but it looks like as written, there is no signature checking for the claims in the passed IDToken - we just go on to stuff the extracted data into the DB, under the newly chosen random string, and return that string. Later auth attempts using the token should succeed, because they look it up in the DB and retrieve the claims that were given. So I could use this to grant myself access to anything.

It is also likely that I haven't understood the entire dex interaction, and that there is some place where the authenticity of the dex payload is checked, or that there's a nuance I'm missing about how this handler is exposed.

Hoping that you can enlighten me :-)

Reply:
The `/sky/issuer/token` endpoint (and the OAuth2 token endpoint in general) supports several different flows, some of which require previous "hops" to get there (e.g. logging in via GitHub using the `authorization_code` grant type). For each of the possible flows, Dex validates that the flow was "successful" (whatever that means for the particular grant type - e.g. correct password for the `password` grant, valid auth code for the `authorization_code` grant, etc). None of the grant types, however, support providing your own claims - Dex tells you what your claims are.

My point being, Dex is still doing all the validation to determine whether the request to `/sky/issuer/token` is valid, and won't grant a token otherwise.

This new middleware delegates to Dex, parses and stores the claims without validation, and sends back a modified response. I didn't see any value in validating the claims against Dex's signing keys for the following reasons:

1. We aren't delegating to an unknown third-party - this is a piece of code that's bundled into the Concourse binary itself
2. We aren't hitting a network interface that could be tampered with - the Dex handlers and request/response handoff is all done in memory

That said, if you think there's still a vulnerability here, I'd be happy to hear you out. I'm far from a security expert, so I could be wrong on this.

Reply:
Aha, I think I did not understand the "delegation to dex" part - that's the `httptest.NewRecorder` business where it has the inner `ServeHTTP`? So this ends up just tweaking the response from Dex, and Dex still does whatever it does to authenticate the total flow.

Reply:
Yep, exactly